File: python-service/routers/sensors.py
"""Sensor CRUD endpoints."""

import logging
import uuid

from fastapi import APIRouter, Depends, HTTPException, Request, Response, status

from middleware.auth import verify_token

logger = logging.getLogger("sensor_service")
from models.sensor import Sensor, SensorCreate, SensorList, SensorUpdate
from repositories.sensor_repository import SensorRepository, get_sensor_repository

router = APIRouter(prefix="/sensors", tags=["sensors"])


@router.get("", response_model=SensorList)
def list_sensors(
    repo: SensorRepository = Depends(get_sensor_repository),
    _: str = Depends(verify_token),
):
    """
    List all sensors.

    Returns all sensors in the database with a count.
    """
    sensors = repo.get_all()
    return SensorList(sensors=sensors, count=len(sensors))


@router.get("/{sensor_id}", response_model=Sensor)
def get_sensor(
    sensor_id: str,
    repo: SensorRepository = Depends(get_sensor_repository),
    _: str = Depends(verify_token),
):
    """
    Get a sensor by ID.

    Returns the sensor details or 404 if not found.
    """
    sensor = repo.get_by_id(sensor_id)
    if not sensor:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No sensor with id '{sensor_id}'",
        )
    return sensor


@router.post("", response_model=Sensor, status_code=status.HTTP_201_CREATED)
def create_sensor(
    sensor: SensorCreate,
    repo: SensorRepository = Depends(get_sensor_repository),
    _: str = Depends(verify_token),
):
    """
    Create a new sensor.

    Generates a unique ID and sets timestamps automatically.
    """
    created = repo.create(sensor)
    return created


@router.put("/{sensor_id}", response_model=Sensor)
def update_sensor(
    sensor_id: str,
    updates: SensorUpdate,
    request: Request,
    repo: SensorRepository = Depends(get_sensor_repository),
    _: str = Depends(verify_token),
):
    """
    Update an existing sensor.

    Only the provided fields will be updated.
    Publishes a sensor.updated event to RabbitMQ on success.
    Returns 404 if the sensor doesn't exist.
    """
    updated = repo.update(sensor_id, updates)
    if not updated:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No sensor with id '{sensor_id}'",
        )

    # Generate trace ID to follow this event through the pipeline
    trace_id = str(uuid.uuid4())
    logger.info("Sensor updated, publishing event", extra={"sensor_id": updated.id, "trace_id": trace_id})

    # Publish sensor.updated event to RabbitMQ
    publisher = request.app.state.publisher
    publisher.publish_sensor_updated(
        sensor_id=updated.id,
        value=updated.value,
        sensor_type=updated.type.value,
        unit=updated.unit.value,
        trace_id=trace_id,
    )

    return updated


@router.delete("/{sensor_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_sensor(
    sensor_id: str,
    repo: SensorRepository = Depends(get_sensor_repository),
    _: str = Depends(verify_token),
):
    """
    Delete a sensor.

    Returns 204 on success, 404 if not found.
    """
    deleted = repo.delete(sensor_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No sensor with id '{sensor_id}'",
        )
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: go-service/handlers/sensors.go
package handlers

import (
	"database/sql"
	"log/slog"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"iot-sensor-service/messaging"
	"iot-sensor-service/models"
	"iot-sensor-service/repositories"
)

// SensorHandler handles sensor CRUD operations.
type SensorHandler struct {
	repo      repositories.SensorRepository
	publisher *messaging.EventPublisher
}

// NewSensorHandler creates a new sensor handler with the given repository and event publisher.
func NewSensorHandler(repo repositories.SensorRepository, publisher *messaging.EventPublisher) *SensorHandler {
	return &SensorHandler{repo: repo, publisher: publisher}
}

// ListSensors returns all sensors.
func (h *SensorHandler) ListSensors(c *gin.Context) {
	sensors, err := h.repo.GetAll()
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{
			Detail: "Failed to retrieve sensors",
		})
		return
	}

	c.JSON(http.StatusOK, models.SensorList{
		Sensors: sensors,
		Count:   len(sensors),
	})
}

// GetSensor returns a single sensor by ID.
func (h *SensorHandler) GetSensor(c *gin.Context) {
	sensorID := c.Param("id")

	sensor, err := h.repo.GetByID(sensorID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{
			Detail: "Failed to retrieve sensor",
		})
		return
	}

	if sensor == nil {
		c.JSON(http.StatusNotFound, models.ErrorResponse{
			Detail: "No sensor with id '" + sensorID + "'",
		})
		return
	}

	c.JSON(http.StatusOK, sensor)
}

// CreateSensor creates a new sensor.
func (h *SensorHandler) CreateSensor(c *gin.Context) {
	var input models.SensorCreate
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{
			Detail: "Invalid request body: " + err.Error(),
		})
		return
	}

	sensor, err := h.repo.Create(&input)
	if err != nil {
		// Validation errors from the model layer are safe to return
		if _, ok := err.(*models.ValidationError); ok {
			c.JSON(http.StatusBadRequest, models.ErrorResponse{
				Detail: err.Error(),
			})
			return
		}
		slog.Error("Failed to create sensor", "error", err)
		c.JSON(http.StatusServiceUnavailable, models.ErrorResponse{
			Detail: "Service temporarily unavailable",
		})
		return
	}

	c.JSON(http.StatusCreated, sensor)
}

// UpdateSensor updates an existing sensor.
func (h *SensorHandler) UpdateSensor(c *gin.Context) {
	sensorID := c.Param("id")

	var input models.SensorUpdate
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{
			Detail: "Invalid request body: " + err.Error(),
		})
		return
	}

	sensor, err := h.repo.Update(sensorID, &input)
	if err != nil {
		if _, ok := err.(*models.ValidationError); ok {
			c.JSON(http.StatusBadRequest, models.ErrorResponse{
				Detail: err.Error(),
			})
			return
		}
		slog.Error("Failed to update sensor", "error", err, "sensor_id", sensorID)
		c.JSON(http.StatusServiceUnavailable, models.ErrorResponse{
			Detail: "Service temporarily unavailable",
		})
		return
	}

	if sensor == nil {
		c.JSON(http.StatusNotFound, models.ErrorResponse{
			Detail: "No sensor with id '" + sensorID + "'",
		})
		return
	}

	// Generate trace ID to follow this event through the pipeline
	traceID := uuid.New().String()
	slog.Info("Sensor updated, publishing event", "sensor_id", sensor.ID, "trace_id", traceID)

	// Publish sensor.updated event for alert services to consume
	if h.publisher != nil {
		go h.publisher.PublishSensorUpdated(sensor.ID, sensor.Value, sensor.Type, sensor.Unit, traceID)
	}

	c.JSON(http.StatusOK, sensor)
}

// DeleteSensor removes a sensor.
func (h *SensorHandler) DeleteSensor(c *gin.Context) {
	sensorID := c.Param("id")

	err := h.repo.Delete(sensorID)
	if err != nil {
		if err == sql.ErrNoRows {
			c.JSON(http.StatusNotFound, models.ErrorResponse{
				Detail: "No sensor with id '" + sensorID + "'",
			})
			return
		}
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{
			Detail: "Failed to delete sensor",
		})
		return
	}

	c.Status(http.StatusNoContent)
}
